pkg/tfbridge/walk.go

```diff
@@ -168,3 +265,347 @@ func LookupSchemaInfoPath(schemaPath SchemaPath, schemaInfo *SchemaInfo) *Schema
 		return nil
 	}
 }
+
+// Drill down a path from a SchemaInfo object and find a matching SchemaInfo. If no
+// SchemaInfo exists, it will be created.
+func getSchemaInfoPath(schemaPath SchemaPath, schemaInfo *SchemaInfo) *SchemaInfo {
+	if len(schemaPath) == 0 {
+		return schemaInfo
+	}
+
+	switch first := schemaPath[0].(type) {
+	case walk.ElementStep:
+		return getSchemaInfoPath(schemaPath[1:], ensureField(&schemaInfo.Elem))
+	case walk.GetAttrStep:
+		return getSchemaInfoPath(schemaPath[1:], ensureMapKey(&schemaInfo.Fields, first.Name))
+	default:
+		contract.Failf("Unable to walk SchemaPath segment of type %T", schemaPath[0])
+		return nil
+	}
+}
+
+type VisitInfo struct {
+	Root VisitRoot
+
+	shimSchema    shim.Schema
+	getSchemaInfo func() *SchemaInfo
+
+	getShimSchemaMap func() shim.SchemaMap
+	getSchemaInfoMap func() map[string]*SchemaInfo
+}
+
+type VisitResult struct {
+	// If the visitor has done something.
+	//
+	// Visits will only be replayed during runtime startup if `HasEffect: true`.
+	HasEffect bool
+}
+
+func (v *VisitInfo) ShimSchema() shim.Schema                        { return v.shimSchema }
```

Comment:
This is a confusing API, I think. I'm confused just trying to verify this :)

The EnclosingSchemaShimMap is not enclosing anything. It calls getShimSchemaMap which unpacks the resource fields at the current schema location.

Reply:
Ah I think this works fine but I got confused because the implementation assumes this nasty trick that apparently I put in:

```
func wrapSchemaMap(sm shim.SchemaMap) shim.Schema {
	return (&schema.Schema{
		Type: shim.TypeMap,
		Elem: (&schema.Resource{Schema: sm}).Shim(),
	}).Shim()
}
```

This trick unifies SchemaMap and Schema types but it creates Schema values that don't "naturally" occur.  However this is part of the machinery here.. so the resources's properties for example can be unpacked via Elem().(*Resource) but only when it got allocated by this util function.